pkg/controllers/provisioningv2/rke2/machineprovision/controller.go

```diff
@@ -220,16 +252,116 @@ func (h *handler) OnRemove(_ string, obj runtime.Object) (runtime.Object, error)
+			removedNodeNameAnnotation := "etcd." + v2provruntime.GetRuntimeCommand(cluster.Spec.KubernetesVersion) + ".cattle.io/removed-node-name"
+
+			machine, err := h.getMachine(obj)
+
+			if err != nil {
+				return obj, err
+			}
+
+			if machine.Status.NodeRef == nil {
+				// Machine noderef is nil, we should just allow deletion.
+				logrus.Debugf("[MachineProvision] There was no associated K8s node with this etcd dynamicmachine %s. proceeding with deletion", key)
+				return h.doRemove(obj)
+			}
+
+			logrus.Debugf("[MachineProvision] Retrieving node %s from K8s for the dynamic machine %s", machine.Status.NodeRef.Name, key)
+
+			node, err := clientset.CoreV1().Nodes().Get(context.TODO(), machine.Status.NodeRef.Name, metav1.GetOptions{})
+			if err != nil {
+				if apierror.IsNotFound(err) {
+					logrus.Debugf("[MachineProvision] Node %s for dynamic machine %s was not found. proceeding with deletion", machine.Status.NodeRef.Name, key)
+					return h.doRemove(obj)
+				}
+				return obj, err
+			}
+
+			if val, ok := node.Annotations[removeAnnotation]; ok {
+				// check val to see if it's true, if not, continue
+				if val == "true" {
+					// check the status of the removal
+					logrus.Debugf("[MachineProvision] etcd member removal is currently in progress per the annotation %s", removeAnnotation)
+					if removedNodeName, ok := node.Annotations[removedNodeNameAnnotation]; ok {
+						// There is the possibility the annotation is defined, but empty.
+						if removedNodeName != "" {
+							logrus.Infof("[MachineProvision] Encountered empty value in removed node name annotation on dynamic machine %s, proceeding with deletion", key)
+							return h.doRemove(obj)
+						}
+					}
+					return obj, fmt.Errorf("remove annotation already set, waiting for node removal to be successful")
+				}
+			}
+			// The remove annotation has not been set to true, so we'll go ahead and set it on the node.
+			err = retry.RetryOnConflict(retry.DefaultRetry,
+				func() error {
+					node.Annotations[removeAnnotation] = "true"
+					node, err = clientset.CoreV1().Nodes().Update(context.TODO(), node, metav1.UpdateOptions{})
+					return err
+				})
+			if err != nil {
+				// there was an error updating the node
+				return obj, err
+			}
+
+			return obj, fmt.Errorf("waiting for etcd member removal")
+		}
+	}
+	return h.doRemove(obj)
+}
+
+func (h *handler) doRemove(obj runtime.Object) (runtime.Object, error) {
+	logrus.Infof("MachineProvision doing removal now")
+	d, err := data.Convert(obj.DeepCopyObject())
+	if err != nil {
+		return nil, err
+	}
+
 	infraName := d.String("metadata", "name")
 	if !d.Bool("status", "jobComplete") && d.String("status", "failureReason") == "" {
 		return obj, fmt.Errorf("cannot delete machine %s because create job has not finished", infraName)
 	}
 
+	logrus.Infof("MachineProvision Running the removal via run")
```

Comment:
I would advocate for removing this log message or adding more information to it (like machine name).